presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlattenedStreamReader.java

```diff
@@ -0,0 +1,383 @@
+        // We will use the offsetVector as the buffer to read the length values from lengthStream,
+        // and the length values will be converted in-place to an offset vector.
+        boolean[][] inMapVectors = new boolean[inMapStreams.size()][nextBatchSize];
+        boolean[] nullVector = null;
+
+        if (presentStream == null) {
+            for (int i = 0; i < inMapStreams.size(); i++) {
+                inMapStreams.get(i).getSetBits(nextBatchSize, inMapVectors[i]);
+            }
+        }
+        else {
+            nullVector = new boolean[nextBatchSize];
+            int nullValues = presentStream.getUnsetBits(nextBatchSize, nullVector);
+            if (nullValues != nextBatchSize) {
+                for (int i = 0; i < inMapStreams.size(); i++) {
+                    inMapStreams.get(i).getSetBits(nextBatchSize, inMapVectors[i], nullVector);
+                }
+            }
+        }
+
+        MapType mapType = (MapType) type;
+        Type keyType = mapType.getKeyType();
+        Type valueType = mapType.getValueType();
+
+        Block[] valueBlocks = new Block[valueStreamReaders.size()];
+
+        for (int i = 0; i < valueStreamReaders.size(); i++) {
+            int mapsContainingValue = 0;
+
+            for (int j = 0; j < inMapVectors[i].length; j++) {
+                mapsContainingValue += inMapVectors[i][j] ? 1 : 0;
+            }
+
+            if (mapsContainingValue > 0) {
+                StreamReader streamReader = valueStreamReaders.get(i);
+                streamReader.prepareNextRead(mapsContainingValue);
+                valueBlocks[i] = streamReader.readBlock(valueType);
+            }
+            else {
+                valueBlocks[i] = valueType.createBlockBuilder(null, 0).build();
+            }
+        }
+
+        int[] mapOffsets = new int[nextBatchSize + 1];
+        int mapOffsetsIndex = 1;
+        mapOffsets[0] = 0;
+
+        for (int i = 0; i < inMapVectors[0].length; i++) {
+            int mapLength = 0;
+            for (int j = 0; j < inMapVectors.length; j++) {
+                if (inMapVectors[j][i]) {
+                    mapLength++;
+                }
+            }
+
+            mapOffsets[mapOffsetsIndex] = mapOffsets[mapOffsetsIndex - 1] + mapLength;
+            mapOffsetsIndex++;
+        }
+
+        int totalMapEntries = mapOffsets[mapOffsets.length - 1];
+        Block keyBlock = createKeyBlock(keyType, inMapVectors, totalMapEntries);
+        Block valueBlock = createValueBlock(inMapVectors, valueBlocks, totalMapEntries, valueType);
+
+        readOffset = 0;
+        nextBatchSize = 0;
+
+        return mapType.createBlockFromKeyValue(Optional.ofNullable(nullVector), mapOffsets, keyBlock, valueBlock);
+    }
+
+    private Block createValueBlock(boolean[][] inMap, Block[] valueBlocks, int totalMapEntries, Type valueType)
+    {
+        int[] blockPositions = new int[inMap.length];
+        BlockBuilder builder = valueType.createBlockBuilder(null, totalMapEntries);
```

Comment:
@dain : @haozhun , @kevinwilfong  and I have an offline discussion. We note this won't help for structural types. Fixed width types will get good estimation based on entry count anyway.

So it will only help with `VARCHAR`. We feel this might not worth it, especially given `getRetainedSizeInBytes` is not designed for this purpose.